Dalamud/Game/Gui/ContextMenu/ContextMenuGui.cs

```diff
@@ -0,0 +1,501 @@
+
+    private readonly ContextMenuAddressResolver address;
+
+    private readonly Hook<ContextAddonOpenByAgentDelegate> contextAddonOpenByAgentHook;
+    private readonly Hook<ContextMenuOnVf72Delegate> contextMenuOnVf72Hook;
+    private readonly RaptureAtkModuleOpenAddonDelegate raptureAtkModuleOpenAddon;
+
+    [ServiceManager.ServiceConstructor]
+    private ContextMenuGui(TargetSigScanner sigScanner)
+    {
+        this.address = new ContextMenuAddressResolver();
+        this.address.Setup(sigScanner);
+
+        this.contextAddonOpenByAgentHook = Hook<ContextAddonOpenByAgentDelegate>.FromAddress(this.address.ContextAddonOpenByAgent, this.ContextAddonOpenByAgentDetour);
+        this.contextMenuOnVf72Hook = Hook<ContextMenuOnVf72Delegate>.FromAddress(this.address.ContextMenuOnVf72, this.ContextMenuOnVf72Detour);
+        this.raptureAtkModuleOpenAddon = Marshal.GetDelegateForFunctionPointer<RaptureAtkModuleOpenAddonDelegate>(this.address.RaptureAtkModuleOpenAddon);
+
+        this.contextAddonOpenByAgentHook.Enable();
+        this.contextMenuOnVf72Hook.Enable();
+    }
+
+    private unsafe delegate nint ContextAddonOpenByAgentDelegate(RaptureAtkModule* module, byte* addonName, AtkUnitBase* addon, int valueCount, AtkValue* values, AgentInterface* agent, nint a7, ushort parentAddonId);
+    
+    private unsafe delegate bool ContextMenuOnVf72Delegate(AddonContextMenu* addon, int selectedIdx, byte a3);
+    
+    private unsafe delegate nint RaptureAtkModuleOpenAddonDelegate(RaptureAtkModule* a1, uint addonNameId, uint valueCount, AtkValue* values, AgentInterface* parentAgent, long unk, uint ownerAddonId, int unk2);
+
+    /// <inheritdoc/>
+    public event IContextMenu.OnMenuOpenedDelegate OnMenuOpened;
+
+    private Dictionary<ContextMenuType, List<MenuItem>> MenuItems { get; } = new();
+
+    private AgentInterface* SelectedAgent { get; set; }
+
+    private ContextMenuType? SelectedMenuType { get; set; }
+
+    private List<MenuItem>? SelectedItems { get; set; }
+
+    private HashSet<nint> SelectedEventInterfaces { get; } = new();
+
+    private AtkUnitBase* SelectedParentAddon { get; set; }
+
+    // -1 -> -inf: native items
+    // 0 -> inf: selected items
+    private List<int> MenuCallbackIds { get; } = new();
+
+    private IReadOnlyList<MenuItem>? SubmenuItems { get; set; }
+
+    /// <inheritdoc/>
+    public void Dispose()
+    {
+        var manager = RaptureAtkUnitManager.Instance();
+        var menu = manager->GetAddonByName("ContextMenu");
+        var submenu = manager->GetAddonByName("AddonContextSub");
+        if (menu->IsVisible)
+            menu->FireCallbackInt(-1);
+        if (submenu->IsVisible)
+            submenu->FireCallbackInt(-1);
+
+        this.contextAddonOpenByAgentHook.Dispose();
+        this.contextMenuOnVf72Hook.Dispose();
+    }
+
+    /// <inheritdoc/>
+    public void AddMenuItem(ContextMenuType menuType, MenuItem item)
+    {
```

Comment:
Should menu modification functions be only called from the UI thread? Use `ThreadSafety.AssertMainThread()` in that case; otherwise, use thread-safe containers or locks.